pkg/runtime/framework/plugins/jobset/jobset.go

```diff
@@ -87,10 +86,47 @@ func (j *JobSet) ReconcilerBuilders() []runtime.ReconcilerBuilder {
 	}
 }
 
+func (j *JobSet) IdentifyPodNetwork(info *runtime.Info, trainJob *trainer.TrainJob) error {
+	if info == nil || trainJob == nil {
+		return nil
+	}
+	spec, ok := runtime.TemplateSpecApply[jobsetv1alpha2ac.JobSetSpecApplyConfiguration](info)
+	if !ok {
+		return nil
+	}
+	subDomain := trainJob.Name
+	if jobSetNet := spec.Network; jobSetNet != nil && jobSetNet.Subdomain != nil {
+		subDomain = *jobSetNet.Subdomain
+	}
+	for rJobIdx, rJob := range spec.ReplicatedJobs {
+		// TODO: Support multiple replicas for replicated Jobs.
+		// REF: https://github.com/kubeflow/trainer/issues/2318
+		podCount := info.TemplateSpec.PodSets[rJobIdx].CountForNonTrainer
+		if *rJob.Name == constants.JobTrainerNode {
+			podCount = info.RuntimePolicy.MLPolicy.NumNodes
+		}
+		rJobReplicas := 1
+		info.TemplateSpec.PodSets[rJobIdx].Endpoints = func(yield func(string) bool) {
+			for podIdx := range ptr.Deref(podCount, 1) {
+				endpoint := fmt.Sprintf("%s-%s-%d-%d.%s", trainJob.Name, *rJob.Name, rJobReplicas-1, podIdx, subDomain)
```

Comment:
For the MPI use-case, do we need separate endpoints for Launcher and Trainer jobs ?

Reply:
We need to record all Node (=Pod) endpoints to hostfile.

Reply:
Should we exclude the Initializer Job from the PodNetwork ?

Reply:
> Should we exclude the Initializer Job from the PodNetwork ?

It would be better to avoid distinguishing between Initializer and others when constructing the `Info` object since, from outside of the JobSet plugin's POV, they can not understand why `Info.PodSet.Endpoints` is missing in the case of some roles.

Reply:
Yeah, I think that makes a lot of sense! And we only iterate over Launcher + Trainer endpoints in the MPI plugin.

Reply:
Yes
Additionally, `PodSet.Endpoints` is an iterator. So, we can reduce time complexity to calculate the endpoint in case of large Jobs.